api-example.xml

```diff
@@ -1,149 +1,219 @@
+    global custom_checks
+    global success_message
+    answer = json.loads(json.loads(ans)['answer'])
+    grader = Grader(answer, custom_checks, success_message)
+    return grader.grade()
+
+
+## Built-in check functions
+
+def check_presence(check, vectors):
+    if check['vector'] not in vectors:
+        return 'You need to use the {} vector.'.format(check['vector'])
+
+def check_tail(check, vectors):
+    vec = vectors[check['vector']]
+    tolerance = check.get('tolerance', 1.0)
+    expected = check['expected']
+    dist = math.hypot(expected[0] - vec.tail.x, expected[1] - vec.tail.y)
+    if dist > tolerance:
+        return 'Vector {} does not start at correct point.'.format(vec.name)
+
+def check_tip(check, vectors):
+    vec = vectors[check['vector']]
+    tolerance = check.get('tolerance', 1.0)
+    expected = check['expected']
+    dist = math.hypot(expected[0] - vec.tip.x, expected[1] - vec.tip.y)
+    if dist > tolerance:
+        return 'Vector {} does not end at correct point.'.format(vec.name)
+
+def _check_coordinate(check, coord):
+    tolerance = check.get('tolerance', 1.0)
+    expected = check['expected']
+    return abs(expected - coord) > tolerance
+
+def check_tail_x(check, vectors):
+    vec = vectors[check['vector']]
+    if _check_coordinate(check, vec.tail.x):
+        return 'Vector {} does not start at correct point.'.format(vec.name)
+
+def check_tail_y(check, vectors):
+    vec = vectors[check['vector']]
+    if _check_coordinate(check, vec.tail.y):
+        return 'Vector {} does not start at correct point.'.format(vec.name)
+
+def check_tip_x(check, vectors):
+    vec = vectors[check['vector']]
+    if _check_coordinate(check, vec.tip.x):
+        return 'Vector {} does not end at correct point.'.format(vec.name)
+
+def check_tip_y(check, vectors):
+    vec = vectors[check['vector']]
+    if _check_coordinate(check, vec.tip.y):
+        return 'Vector {} does not end at correct point.'.format(vec.name)
+
+def _check_coords(vec, expected, tolerance):
+    for expected_coords, vec_coords in ([expected[0], vec.tail], [expected[1], vec.tip]):
+        if expected_coords[0] == '_':
+            if expected_coords[1] != '_':
+                # Only y given; not interested in x.
+                if abs(expected_coords[1] - vec_coords.y) > tolerance:
+                    return True
+        elif expected_coords[1] == '_':
+            # Only x given; not interested in y.
+            if abs(expected_coords[0] - vec_coords.x) > tolerance:
+                return True
+        else:
+            # Both coordinates given; check distance from expected point.
+            x = expected_coords[0] - vec_coords.x
+            y = expected_coords[1] - vec_coords.y
+            if math.hypot(x, y) > tolerance:
+                return True
+    return False
+
+def check_coords(check, vectors):
+    vec = vectors[check['vector']]
+    expected = check['expected']
+    tolerance = check.get('tolerance', 1.0)
```

Comment:
@e-kolpakov When you define the tolerance in the `expected_result` setting, you already have to pass it as `coord_tolerance`. Same goes for all other checks (`angle_tolerance`, `tail_tolerance`, `length_tolerance`...). They're all converted to just `tolerance` inside `vectordraw.js` code when converting the `expected_result` entry to a list of checks.